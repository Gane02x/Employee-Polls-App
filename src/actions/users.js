export const GET_USERS = "GET_USERS";
export const UPDATE_USERS_QUESTIONS = "UPDATE_USERS_QUESTIONS";
export const UPDATE_USERS_ANSWERS = "UPDATE_USERS_ANSWERS";

export function getUsers(users) {
  return {
    type: GET_USERS,
    users,
  };
}

export function updateUsersQuestions(question) {
  return {
    type: UPDATE_USERS_QUESTIONS,
    authedUser: question.author,
    id: question.id,
  };
}

export function updateUsersAnswers({ authedUser, qid, answer }) {
  return {
    type: UPDATE_USERS_ANSWERS,
    authedUser,
    qid,
    answer,
  };
}
