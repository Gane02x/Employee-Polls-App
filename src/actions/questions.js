import { _saveQuestion, _saveQuestionAnswer } from "../utils/_DATA";
import { updateUsersQuestions } from "./users";

export const GET_QUESTIONS = "GET_QUESTIONS";
export const ANSWER_QUESTION = "ANSWER_QUESTION";
export const SAVE_QUESTION = "SAVE_QUESTION";

export function getQuestions(questions) {
  return {
    type: GET_QUESTIONS,
    questions,
  };
}

export function handleAnswerQuestion({ authedUser, qid, answer }) {
  return (dispatch) => {
    return _saveQuestionAnswer({ authedUser, qid, answer }).then(() => {
      dispatch(addAnswer({ authedUser, qid, answer }));
    });
  };
}

export function handleSaveQuestion(question) {
  return (dispatch) => {
    return _saveQuestion(question)
      .then((formattedQuestion) => {
        dispatch(saveNewQuestion(formattedQuestion));
        dispatch(updateUsersQuestions(formattedQuestion));
      })
      .catch((e) => console.log("Error at handleSaveQuestion:", e));
  };
}
function addAnswer({ authedUser, qid, answer }) {
  return {
    type: ANSWER_QUESTION,
    authedUser,
    qid,
    answer,
  };
}
function saveNewQuestion(question) {
  return {
    type: SAVE_QUESTION,
    question,
  };
}
