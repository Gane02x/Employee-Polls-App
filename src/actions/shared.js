import { getUsers } from "./users";
import { getQuestions } from "./questions";
import { setAuthedUser } from "./authedUser";
import { _getQuestions, _getUsers } from "../utils/_DATA";

export function getInitialData() {
  return Promise.all([_getUsers(), _getQuestions()]).then(
    ([users, questions]) => ({
      users,
      questions,
    })
  );
}

export function handleInitialData() {
  return (dispatch) => {
    return getInitialData().then(({ users, questions }) => {
      dispatch(getUsers(users));
      dispatch(getQuestions(questions));
    });
  };
}

export function handleLogin(aid) {
  return (dispatch) => {
    dispatch(setAuthedUser(aid));
  };
}
