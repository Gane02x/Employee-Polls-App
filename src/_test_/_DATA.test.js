import { getInitialData } from "../actions/shared";
import {
  generateUID,
  _saveQuestion,
  _saveQuestionAnswer,
} from "../utils/_DATA";
import { formatDate } from "../utils/api";

// An async unit test to verify that every existing user is returned
test("returns users and questions", async () => {
  const { users, questions } = await getInitialData();

  const numUsers = Object.keys(users).length;
  const numQuestions = Object.keys(questions).length;
  expect(numUsers).toEqual(4);
  expect(numQuestions).toEqual(6);
});

// A snapshot test to verify that the date is formatted
test("returns a correctly formatted date to the page", () => {
  expect(formatDate(new Date("January 1, 2000"))).toMatchSnapshot();
});

// An unit test to verify that a unique id is returned
test("returning an unique id", () => {
  const id1 = generateUID();
  const id2 = generateUID();
  const id3 = generateUID();
  expect(id1).not.toEqual(id2);
  expect(id1).not.toEqual(id3);
  expect(id2).not.toEqual(id3);
});

// UNIT TESTS FOR _saveQuestion()

// An async unit test to verify that the saved question is returned
test("returning true when correctly formatted data is passed into _saveQuestion()", async () => {
  const questionExample = {
    author: "name",
    optionOneText: "1",
    optionTwoText: "2",
  };

  const submit = await _saveQuestion(questionExample);

  const { author, optionOneText, optionTwoText } = submit;

  expect(author).toEqual("name");
  expect(submit.optionOneText).toEqual(optionOneText);
  expect(submit.optionTwoText).toEqual(optionTwoText);
});

// An async unit test to verify that an error is returned if incorrect data is passed to the function
test("returns an error if incorrect data is passed to _saveQuestion()", async () => {
  const questionExample = {
    author: null,
    optionOneText: null,
    optionTwoText: null,
  };
  await expect(_saveQuestion(questionExample)).rejects.toBe(
    "Please provide optionOneText, optionTwoText, and author"
  );
});

// UNIT TESTS FOR _saveQuestionAnswer()

// An async unit test to verify that true is returned when correctly formatted data is passed to the function
test("returning true when correctly formatted data is passed into _saveQuestionAnswer()", async () => {
  const answerExample = {
    authedUser: "sarahedo",
    qid: "8xf0y6ziyjabvozdd253nd",
    answer: "optionTwo",
  };
  const { users, questions } = await _saveQuestionAnswer(answerExample);

  expect(
    users[answerExample.authedUser].answers[answerExample.qid] ===
      answerExample.answer
  ).toBe(true);
  expect(
    questions[answerExample.qid][answerExample.answer].votes.includes(
      answerExample.authedUser
    )
  ).toBe(true);
});

// An async unit test to verify that an error is returned if incorrect data is passed to the function
test("returns an error if incorrect data is passed to _saveQuestionAnswer()", async () => {
  const answerExample = {
    authedUser: null,
    qid: null,
    answer: null,
  };

  await expect(_saveQuestionAnswer(answerExample)).rejects.toBe(
    "Please provide authedUser, qid, and answer"
  );
});
