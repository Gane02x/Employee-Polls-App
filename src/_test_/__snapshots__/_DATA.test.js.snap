// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns a correctly formatted date to the page 1`] = `"12:00AM | 1/1/2000"`;
